README.md
# PLOTIT
## An Art Package for the ZX Spectrum Next

[![PLOTIT Launch Video](http://img.youtube.com/vi/wpi9r7q6J8k/0.jpg)](http://www.youtube.com/watch?v=wpi9r7q6J8k "PLOTIT Launch Video")


* Please report crashing bugs WITH steps to reproduce [here](https://github.com/StalePixels/PLOTIT/issues)

## Latest Version
### v0.8.0 - That *was* a mistake
 * Undo Tool
 * Palette Editor
 * Magnify Preview
 * Settings for Magnify, Line, Box and Circle
 * Public Plugin API
## Previous Versions

### v0.7.9 - Grew a little bit
 * Tooltips and rewritten HUD, now fully Core v3 Compatible.

### v0.7.1 - Minimum Lovable Product
 * Keyboard Input - Mouse now optional, but still the optimum control method
 
### v0.7.0 - Minimum Lovable Product
 * Three new tools
   * Circle
   * Dropper
   * Stamps
 * Image saving(!) 
 * SL2 support
 * THREE hidden easter eggs
 * Quick Read & Write for files, enabling proper emulator use

### v0.6.7 - #BlameLampros
 * Initial *EXPERIMENTAL* CSpect Support
 * Initial Keyboard Support

### v0.6.6 - #MerryNextMas
 * Initial Public Offering
 * No Saving of Files
 * No Plugins
 * Only loads native .nxi files
 * Easter Egg - 'Rubber, not Glue'


docs/CHANGELOG.md
# ChangeLog

PLOTIT is a full-featured image editor, developed across multiple years.

## Controls
PLOTIT is designed to be controlled by a mouse, it also has keyboard shortcuts
which may help speed up some workflows or certain kinds of detail work.

|   **Key**   | **Use**
|-------------|----
|           q | Move mousepointer up one pixel
|           a | Move mousepointer down one pixel
|           o | Move mousepointer left one pixel
|           p | Move mousepointer right one pixel
| CapsShift+Q | Move mousepointer up eight pixels
| CapsShift+A | Move mousepointer down eight pixel
| CapsShift+O | Move mousepointer left eight pixel
| CapsShift+P | Move mousepointer right eight pixel
|           n | Left Mouse Button (Primary Action)
|           m | Right Mouse Button (Secondary Action)

The keyboard controls are intended as a suppliment to the mouse, and never as
a total replacement for the mouse - therefore there's likely many features and
functions that you can only do with the mouse.

## v0.8.1 - That *was* a mistake
### Bug Fixes ###
* Fixed a bug that would only occur if you'd not disabled the autoexec-welcome.bas message. (I know, mad right...) Thanks to Paul Smallman for finding it, the reproduction video, and starting me on the right tracks.
* Fix a bug where some PLUGIN related code was executed when in LITE mode

## v0.8.0 - That *was* a mistake
### New Features
* Undo
* Rewrote Pens and Pencil tools
* Entirely new Menu/HUD complete
* Palette Editor
* Preview when zoomed in
* Beta Plugin API
### Bug Fixes ###
* Fixed several UI rendering issues with stamps and sidebar
* Better handling when switching from config mode to editor modes

## v0.7.9 - Growing just a little bit
### New Features
 * No more Canvas Scrolling
 * Tooltips in lower status bar
### Bug Fixes ###
 * Rewrote mouse driver

## v0.7.1 - Minimum Lovable Product
### New Features ###
 * Keyboard Input - Mouse now optional, but still the optimum control method. Keyboard input allows precise pixel-perfect positioning of your PLOTIT pointer, ideal when used in combination with the mouse.
### Bug Fixes ###
 * Speed ups and Overflow fixes on Fill tool

## v0.7.0 - Minimum Lovable Product
### New Features ###
 * Stand Alone .nex - only the demos used in the "Rogue's Gallery" external to the main app
 * Save your work, both Save, and Save As to keep backups as you go now work! 
 * Freeform Circles (a.k.a. Ellipses) 
 * EyeDropper - select colour from current canvas tool
 * Rubber Stamp tool, with paginated tool settings window
 * Better error handling
 * Corrupt .NXI detection
 * SL2 support
 * Keyboard Hotkey Support
### Bug Fixes ###
 * Sped up Pen Tool, 
 * Sped up Box Tool, especially after multiple uses
 * Left/Right panning arrows no longer backwards!
 * Wierd bug that meant you could not open any more files if you opened one whose filename was 8characters long!
 * Identified lots more still useful debugging to remove from release builds in future too for more speeds ups
 * Zoom Preview no longer corrupts main canvas
 * Fix crash where clicking between tools in Tool Settings window could crash app
 * Fix bug where opening file after viewing credits would sometimes crash app
 * Fix fill bug where paint would overflow under toolbar if window was panned hard left
### Developer Features ###
 * Single Fire Plugin API v0  
 
## v0.7.0
 * Internal Developer Release

### v0.6.7 - Lampros Made Me Do It
### New Features ###
 * ESXDOS file support

## v0.6.6 - #MerryNextmas
### New Features ###
 * All of them
### Bug Fixes ###
 * None

## v0.6.0 
 * Internal Developer Release
 * Initial closed alpha test


dev/plugins/make.sh
#!/bin/bash
#  VERSION 1
#
#  DEV NOTES
#
#   Current BITMASK for entrypoint types
#     BIT     WHEN SET
#      1       Instant Execute
#               Plugin will execute as soon as selected
#      2       Execute on Mouse Click
#               Menu will close, plug in execute on next
#               mouseclick
#      3
#      4
#      5
#      6
#      7
#      8       Writes Banks 17->24 (128k total) to SD
#               and reloads original data after plugin
#               exits.

#   Current BITMASK for pre-flight options
#     BIT     WHEN SET
#      1       Create undo restore point before execute
#      2
#      3
#      4
#      5
#      6
#      7
#      8
#   See dev/docs/api.txt for more details

## A Shell script to turn a .BIN file into a .PLP file
#
#  A .PLP file is an exported memorybank, starting at 0xC000, and ending at most 0xFFFF
#   prepended with a special header to tell PLOTIT how to handle and display the menu

##
#
# Plugin headers are exactly 44bytes long, including the entry count, and menu name
#  *PLUS* an additional 18bytes per entry point
# Plugin bodies are 16384 bytes long (the below script takes care of padding any
#  that is required during assembly
# Plugins *ARE* checked for a size-match against their Entry Count in the plugin
#  header - i.e. a valid PLP file is *at least* 16446 - malformed plugins are rejected

# Create a new blank file, exactly 16k long. We cheat here, and use a premade one - exactly 16k of NULL bytes...
cp util/blank16k.bin /tmp/test.plotit_plugin_bin

# Replace the NULL bytes with the output of our compile -- this handles any insances where we're too short with padding
dd conv=notrunc if=bank.bin of=/tmp/test.plotit_plugin_bin

# PLOTIT VERSION CHECK HEADER - RESERVED (DO NOT CHANGE)
echo -n -e "PLOTIT-PLUGIN-v1\x00" > test.plotit_plugin_header

# PLUGIN MENU TITLE - UCHAR[16] Fixed length string - menu title
echo -n -e "Jim Was 'Ere    " >> test.plotit_plugin_header

# PLUGIN VERSION MAJOR - UINT8
echo -n -e "\x00" >> test.plotit_plugin_header

# PLUGIN VERSION MINOR - UINT8
echo -n -e "\x01" >> test.plotit_plugin_header

# PLUGIN VERSION HASH - UCHAR[8]
echo -n -e "ANY HASH" >> test.plotit_plugin_header

# PLUGIN ENTRY COUNT - UINT8
echo -n -e "\x04" >> test.plotit_plugin_header                           # Entry Count in this plugin (4)


## PLUGIN ENTRY LOOP STARTS NOW #1
### PLUGIN DEFINITION FLAGS - UINT8
echo -n -e "\x01"               >> test.plotit_plugin_header             # INSTANT EXECUTE
### PLUGIN PREFLIGHT FLAGS - UINT8
echo -n -e "\x00"               >> test.plotit_plugin_header             # No undo created before execute
### PLUGIN ENTRY NAME = UCHAR[16]
echo -n -e "Invert Now      "   >> test.plotit_plugin_header             # String, fixed length, no terminator
## PLUGIN ENTRY LOOP ENDS NOW #2


## PLUGIN ENTRY LOOP STARTS NOW #2
### PLUGIN DEFINITION FLAGS - UINT8
echo -n -e "\x01"               >> test.plotit_plugin_header             # INSTANT EXECUTE
### PLUGIN PREFLIGHT FLAGS - UINT8
echo -n -e "\x01"               >> test.plotit_plugin_header             # Restore point created before execute
### PLUGIN ENTRY NAME = UCHAR[16]
echo -n -e "Pnoid Invert Now"   >> test.plotit_plugin_header             # String, fixed length, no terminator
## PLUGIN ENTRY LOOP ENDS NOW #2


## PLUGIN ENTRY LOOP STARTS NOW #3
### PLUGIN DEFINITION FLAGS - UINT8
echo -n -e "\x02"               >> test.plotit_plugin_header             # CLICK EXECUTE
### PLUGIN PREFLIGHT FLAGS - UINT8
echo -n -e "\x00"               >> test.plotit_plugin_header             # No undo created before execute
### PLUGIN ENTRY NAME = UCHAR[16]
echo -n -e "      Invert L8r"   >> test.plotit_plugin_header             # String, fixed length, no terminator
## PLUGIN ENTRY LOOP ENDS NOW #3


## PLUGIN ENTRY LOOP STARTS NOW #4
### PLUGIN DEFINITION FLAGS - UINT8
echo -n -e "\x02"               >> test.plotit_plugin_header             # CLICK EXECUTE
### PLUGIN PREFLIGHT FLAGS - UINT8
echo -n -e "\x01"               >> test.plotit_plugin_header             # Restore point created before execute
### PLUGIN ENTRY NAME = UCHAR[16]
echo -n -e "Pnoid Invert L8r"   >> test.plotit_plugin_header             # String, fixed length, no terminator
## PLUGIN ENTRY LOOP ENDS NOW #4


cat test.plotit_plugin_header > test.plp
cat /tmp/test.plotit_plugin_bin >> test.plp
rm /tmp/test.plotit_plugin_bin




dev/plugins/incs/plotit.asm
PUBLIC _ula_display
defc _ula_display = 0x4000

PUBLIC _ula_attributes
defc _ula_attributes = 0x5800

PUBLIC _sysvars
defc _sysvars = 0x5b00

PUBLIC _tiles
defc _tiles = 0x5c00

PUBLIC _tilemap
defc _tilemap = 0x6000

PUBLIC _fgColor
defc _fgColor = 0x6A00

PUBLIC _bgColor
defc _bgColor = 0x6A01

PUBLIC _plotit_core_state
defc _plotit_core_state = 0x6A02

PUBLIC _toolRegisters;
defc _toolRegisters = 0x6A03    ; 32bytes

PUBLIC _activeTool;
defc _activeTool = 0x6A24       ; uInt 8

PUBLIC _mouseBtns;
defc _mouseBtns = 0x6A25        ; uInt 8

PUBLIC _mouseX;
defc _mouseX = 0x6A26           ; uInt16

PUBLIC _mouseY;
defc _mouseY = 0x6A28           ; uInt16

PUBLIC _keyboard;
defc _keyboard = 0x6A2A         ; unsigned char

PUBLIC _canvasX;
defc _canvasX = 0x6A2B          ; uInt8

PUBLIC _canvasY;
defc _canvasY = 0x6A2C          ; uInt8

PUBLIC _plotColour;
defc _plotColour = 0x6A2D       ; uInt8

PUBLIC _plotTarget;
defc _plotTarget = 0x6A2E       ; uInt8

PUBLIC _undoMaxLevels;
defc _undoMaxLevels = 0x6A2F         ; uInt 8

PUBLIC _visible_screen_MMU;
defc _visible_screen_MMU = 0x6A30

PUBLIC _backing_store_MMU;
defc _backing_store_MMU = 0x6A36

PUBLIC _scratch_screen_MMU;
defc _scratch_screen_MMU = 0x6A40

PUBLIC _undo_screen_MMU;
defc _undo_screen_MMU = 0x6A46

PUBLIC _toolMemory;
defc _toolMemory = 0x6A50

PUBLIC _mouseCX;
defc _mouseCX = 0x6A70         ; uInt 8

PUBLIC _mouseCY;
defc _mouseCY = 0x6A71         ; uInt 8

PUBLIC _pluginActiveID;
defc _pluginActiveID = 0x6A72         ; uInt 8

PUBLIC _pluginEntryPoint;
defc _pluginEntryPoint = 0x6A73         ; uInt 8

PUBLIC _scratchUser;
defc _scratchUser = 0x6A74         ; uInt 8

PUBLIC _dst_sfn;
defc _dst_sfn = 0x6A75              ; char[13] (null terminated)

PUBLIC _palette_flags;
defc _palette_flags = 0x6A83   ; bitfield
                               ; Bit:      0 1 2 3 4 5 6 7
                               ;           |
                               ; Modified: +

; 0x6A84 - 0x61FF blank

PUBLIC _paletteMemory;
defc _paletteMemory = 0x6B00

PUBLIC _buf_256;
defc _buf_256 = 0x6D00

; 0x6E00 next...


dev/plugins/example/invertScreen.asm
; This began life as the same code in the "plugintest" folder, by Jim Bagley - please see there for his
;  original version - here I've done nothing more than tweak the memory checks, and port it to sjasmplus
;
;           build with
;                   sjasmplus invertScreen.asm
;           will result in a "bank.bin", exactly 16k in size
;           this file is suitable for use with make.sh for adding headers;
;
;               -Dx 2021/01/08

                DEVICE ZXSPECTRUMNEXT
                ORG $c000

                push hl                     ; All these pushes are paranoia, and not required
                push de                     ; but this "boilerplate" is useful for other plugin
                push bc                     ; authors.   This "short and sweet" version is that
                push af                     ; you should back up, and restore, anything you change
                push ix                     ; that's not explicitly named as "changable" or
                push iy                     ; restored at exit by the plugin docs.
                exx                         ;
                push hl                     ; And, yes, that includes MMUs, SysVars, and
                push de                     ; the registers - *ALL* the registers, if
                push bc                     ; that turns out to be something you're changing...
                ex af,af                    ;
                exx                         ;
                push af                     ; And (see the bottom for details) remember to put
                ex af,af                    ;  everything back again

                ld d,9*2                    ;9 is our VRAM MMU

LoopOuter:	    ld a,d                      ; load the MMU into the accumulator
                nextreg $57,a               ; Technically this is our plugin memoryspace, but we have no
                ld hl,$e000                 ; code up here - if we did 4&5 contain only the core app code

LoopInner:	    ld a,(hl)                   ; load the pixel into the accumulator
                cpl                         ;   flip all the bits
                ld (hl),a                   ;   and put the pixel back
                inc l                       ; one down,
                jr nz,LoopInner             ; 48thousand more to go...      Inner Loop Low Byte &
                inc h                       ;                   and then
                jr nz,LoopInner             ;                               High Byte part
                inc d                       ; bump the MMU
                ld a,d                      ;   are we out of VRAM?
                cp 12*2                     ; if we hit MMU 12 we have gone too far...
                jr nz,LoopOuter             ;   Start on next MMU

                ex af,af                    ; Basically undo all the boilerplate from the top.
                pop af                      ;
                ex af,af                    ; And, as mentioned up there, you probably don't
                exx                         ; need all of this - it's just here to both "remind you"
                pop bc                      ; and as to serve as a boilerplate new plugin authors
                pop de                      ; can use and built upon
                pop hl                      ;
                exx                         ; Andand, just like above, include the shadow registers.
                pop iy                      ;
                pop ix                      ;
                pop af                      ;
                pop bc                      ;
                pop de                      ;
                pop hl                      ;

                ret                         ; Andandand, remember to return to PLOTIT when done :-P

                SAVEBIN "bank.bin",$c000,$ffff

dev/plugins/incs/plotit_API.h
//
// Created by D Rimron-Soutter on 30/12/2020.
//


#include <stdint.h>
#include <stdbool.h>
#include "zxnext_layer2.h"

#ifndef PLOTIT_API_H
#define PLOTIT_API_H

#define PLOTIT_MMU_SCRATCH      0x02
#define PLOTIT_MMU_DEFAULT      0x06

#define PLOTIT_STAMP_PATTERN    30

#define ZXNEXTOS_MMU_VRAM1      0x09//09
#define ZXNEXTOS_MMU_VRAM2      0x0A//10
#define ZXNEXTOS_MMU_VRAM3      0x0B//11
#define PLOTIT_MMU_IMAGE_DATA1  0x0C//12
#define PLOTIT_MMU_IMAGE_DATA2  0x0D//13
#define PLOTIT_MMU_IMAGE_DATA3  0x0E//14

#define PLOTIT_MMU_PLUGIN_CODE  0x0F//15

// -- UNDO START
#define PLOTIT_MMU_UNDO_6_1     0x10//16
#define PLOTIT_MMU_UNDO_6_2     0x11//17
#define PLOTIT_MMU_UNDO_6_3     0x12//18

#define PLOTIT_MMU_UNDO_5_1     0x13//19
#define PLOTIT_MMU_UNDO_5_2     0x14//20
#define PLOTIT_MMU_UNDO_5_3     0x15//21

#define PLOTIT_MMU_UNDO_4_1     0x16//22
#define PLOTIT_MMU_UNDO_4_2     0x17//23
#define PLOTIT_MMU_UNDO_4_3     0x18//24

#define PLOTIT_MMU_UNDO_3_1     0x19//25
#define PLOTIT_MMU_UNDO_3_2     0x1A//26
#define PLOTIT_MMU_UNDO_3_3     0x1B//27

#define PLOTIT_MMU_UNDO_2_1     0x1C//28
#define PLOTIT_MMU_UNDO_2_2     0x1D//29
#define PLOTIT_MMU_UNDO_2_3     0x1E//30

#define PLOTIT_MMU_UNDO_1_1     0x1F//31
#define PLOTIT_MMU_UNDO_1_2     0x20//32
#define PLOTIT_MMU_UNDO_1_3     0x21//33
// -- UNDO END

#define PLOTIT_MMU_SCRATCH_1    0x22//34
#define PLOTIT_MMU_SCRATCH_2    0x23//35
#define PLOTIT_MMU_SCRATCH_3    0x24//36

#define PLOTIT_MMU_RESERVED_4   0x25//37
#define PLOTIT_MMU_RESERVED_3   0x26//38

#define PLOTIT_MMU_STAMP_BOOK   0x27//39
#define PLOTIT_MMU_UI_SPRITES   0x28//40

#define PLOTIT_MMU_RESERVED_2   0x29//41
#define PLOTIT_MMU_RESERVED_1   0x2A//42

#define PLOTIT_MMU_PLUGINS      0x2B//43
#define PLOTIT_MMU_CREDITS      0x2C//44
#define PLOTIT_MMU_TOOLS        0x2D//45
#define PLOTIT_MMU_ENGINE       0x2E//46
#define PLOTIT_MMU_UTILS        0x2F//47
#define PLOTIT_MMU_ROM          0xFF//FF

#define MOUSE_BTN_PRIMARY       0x01
#define MOUSE_BTN_SECONDARY     0x02

#define TOOL_PENCIL             0x00
#define TOOL_PEN                0x01
#define TOOL_LINE               0x02
#define TOOL_BOX                0x03
#define TOOL_CIRCLE             0x04
#define TOOL_FILL               0x05
#define TOOL_MAGNIFY            0x06
#define TOOL_EYEDROPPER         0x07
#define TOOL_STAMP              0x08
#define TOOL_NOTREADY           0x09

#define PALETTE_FLAG_MODIFIED   0x01

//#define DEFAULT_DEBOUNCE_MS  400
#define DEFAULT_DEBOUNCE_FRAMES  10

#define DEFAULT_UNDO_LEVELS     6

///////////////////////////////
// PUBLIC MEMORY MAP (See NOTE 0)
///////////////////////////////
/*
 * 0x4000 ULA Display File
 * 0x5800 Attributes                    NOTE 1
 * 0x5b00 System Variables              NOTE 2
 * 0x5c00 System Variables              NOTE 3: re: _tiles[0-31]
 * 0x5d00 _tiles[32-127]
 * 0x6000 _tilemap[32][40]
 * 0x6A00 _fgColor
 * 0x6A01 _bgColor
 * 0x6A02 _plotit_core_state
 * 0x6A03 _toolRegisters[8]             NOTE 6: para 1
 * 0x6A24 _activeTool
 * 0x6A25 _mouseBtns
 * 0x6A26 _mouseX                       NOTE 7
 * 0x6A28 _mouseY                       NOTE 7
 * 0x6A2A _keyboard
 * 0x6A2B _canvasX                      NOTE 7
 * 0x6A2C _canvasY                      NOTE 7
 * 0x6A2D _plotColour
 * 0x6A2E _plotTarget
 * 0x6A2F _undoMaxLevels
 * 0x6A30 _visible_screen_MMU           NOTE 4: area 1
 * 0x6A36 _backing_store_MMU            NOTE 4: area 2
 * 0x6A40 _scratch_screen_MMU           NOTE 4: area 3
 * 0x6A46 _undo_screen_MMU              NOTE 5
 * 0x6A50 _toolMemory[16]               NOTE 6: para 2
 * 0x6A70 _mouseCX
 * 0x6A71 _mouseCY
 * 0x6A72 _pluginActiveID
 * 0x6A73 _pluginEntryPoint
 * 0x6A74 _scratchUser;
 * 0x6175 _dst_sfn;
 * 0x6183 _palette_flags;               NOTE 8:
 *
 * 0x6B00 _paletteMemory
 * 0x6D00 _buf_256
 * 0x6E00
 *
 */

// Mostly here as a conceptual hack / example on how memory addressing works
extern uint8_t __ula_display[32][192];
extern uint8_t __ula_attributes[32][24];

struct __tiles
{
    unsigned char bmp[8];                       // each tile image is 8 bytes (64px@1bpp)
};
extern struct __tiles tiles[128];               // 128 tile images in total

struct __tilemap
{
    unsigned char tile;     // 0-7 = 8 bit tile number

    // Flags match NEXTREG: 0x6C (108) R/W => Default Tilemap Attribute format
    //  bits 7-1 = Palette Offset
    //  bit 0    = ULA over tilemap
    unsigned char flags;    // tile attribute (palette offset, rotation, etc)
};
extern struct __tilemap tilemap[32][40];        // 32 rows, 40 columns in each...

// Painting colours
extern uint8_t fgColor, bgColor;

// App's internal flag state
#define PLOTIT_STATE_DIRTY              1
extern uint8_t plotit_core_state;

// 32bytes of space reserved for tools to operate on when running so they can pass settings to other memory areas
extern int16_t toolRegisters[16];

// And what the current tool is
extern uint8_t activeTool;
// Flag for special handling if action was via menu
extern uint8_t activatedFromMenu;

// User Input - Mouse
extern uint8_t mouseBtns;
extern int16_t mouseX;
extern int16_t mouseY;

// Last key scanned
extern unsigned char keyboard;

extern uint8_t canvasX;
extern uint8_t canvasY;
extern uint8_t plotColour;
extern uint8_t plotTarget;

// How many undo levels the app has enabled (to see if you can steal any)
extern uint8_t undoMaxLevels;

// Some memstructs for handling screen addresses
extern layer2_screen_t visible_screen_MMU;
extern layer2_screen_t backing_store_MMU;
extern layer2_screen_t scratch_screen_MMU;
extern layer2_screen_t undo_screen_MMU;

// Tools persist their selections (e.g., brush selection) - 16 slots, only 9 currently used.
extern int16_t toolMemory[16];

// Some tools pass X&Y coords between each other, plugins can use this for local coord maths/cache
extern uint8_t mouseCX;
extern uint8_t mouseCY;

// Your plugin could be loaded in any order, so when you're executed your order ID will be here (starts at 0)
extern uint8_t pluginActiveID;

// When your plugin is called, this will be an integer marking which entry point (aka menu item)
// in the PLP file triggered it  (starts at 0)
extern uint8_t pluginEntryPoint;

// If the ID in scratchUser matches your pluginActiveID - you can use the scratch memory without backing it up to SD first
// -- backing the scratch memory up to SD isn't forced, but if you don't, make sure you change the scratchUser ID on exit
// so other plugins can tell you've changed it.
// You should leave your own pluginActiveID in this memory location _if_ you have left data here after your plugin
// has finished running -- but be warned, this it's an honour system - bad plugins may never update this upon execution.
// This is probably best used for "start up code" or loading extra assets/binaries into, you can save repeated startup
// time if the 'cache' is "yours" and not important/irreplacable state data.
extern uint8_t scratchUser;

// The currently active pallete as last uploaded into the L2 hardware - here for easy reference, do not edit
extern uint16_t paletteMemory[256];

// Simple 0.25K buffer, used for file ops, UI &  stamp management - you can use it for your plugin, if desired
extern uint8_t buf_256[256];

//  Short file name for the currently active/open document
extern unsigned char dst_sfn[13];

// Palette flags - metadata about current palette state, see NOTE 8
extern unsigned char palette_flags;


/*
 * NOTE 0
 *      The public memory map is yours to do with as you see fit,
 *      but not without consequences.  It is best to consider all of
 *      it essential for the correct operation of PLOTIT, _outside_
 *      of your plugin, and thus - anything you change, you should put
 *      back when done.
 *
 *      You can page the public memory space in and out as you see fit,
 *      and plugins are executed with interrupts disabled for this reason
 *      although PLOTIT expects BANK5 (the public memory) back in the
 *      default location when you return control to the main app.
 *
 *      There are also certain areas of the public memory map that are
 *      used to pass either parameters to your plugin, or for hardware
 *      support, and those are listed below.
 *
 * NOTE 1
 *      The ULA screen is used exclusively by PLOTIT for tool previews.
 *      Default paper is non-bright blue, and using this will make the ULA
 *      appear transparent, allowing you to do similar FX for your plugin.
 *
 * NOTE 2
 *      PLOTIT doesn't use the system variables for itself, except when
 *      the file browser is opened - buy they are preserved intact so that
 *      your plugin has the same OS options available to it (outwith channels)
 *
 * NOTE 3
 *      The tile definitions will be placed at address 0x5c00.
 *      However the ROM source font only defines characters 32 - 127
 *      which is the printable part of the ascii set so we will
 *      only be using those tile definitions stored in addresses
 *      0x5d00 to 0x5fff --Allen
 *
 *      You can replace the tile "entities" from 0 to 31 with your own when creating
 *      custom UIs, but you should to replace the sysvars when done rendering.
 *
 * NOTE 4
 *      The Video memory is broadly split into (public) sections
 *      1) visible_screen_MMU - this is where you can draw, drawing to this is
 *          drawing to the user's canvas.
 *      2) backing_store_MMU - Backing store used during creation of custom user
 *          interfaces - when configured plugins will be launched with a copy of
 *          the canvas already in this memory, making changes to visible_screen_MMU
 *          non-destructive.  Upon exiting the UI of your plugin this memory is
 *          restored over visible_screen_MMU.  You should not draw to backing_store_MMU
 *          directly without understanding the consequences.
 *      3) scratch_screen_MMU - 48k "blank", use as an SD load cache, or area of creating
 *          custom effects, blit space, etc.  Also can be used as extra scratch data
 *          for your plugin, if required.   Cannot be relied on up to be in the same state
 *          as left between plugin executions.  You can write to this as required, there
 *          is no need to back up data before changing it.
 *      The structures of memory in this area is as follows:
 *          Byte 00:    reserved
 *          Byte 01:    reserved
 *          Byte 02:    MMU Bank number, top 16k
 *          Byte 03:    MMU Bank number, middle 16k
 *          Byte 04:    MMU Bank number, bottom 16k
 *          Byte 05:    MMU Bank number, calling bank when using zxnext_layer2 api (deprecated - do not use)
 *
 * NOTE 5
 *      undo_screen_MMU is the state of the previous undo step - this permits you to perform related effects between
 *      an image, and the previous state, in a plugin if required.  Etiquette says "don't edit this" in case the user
 *      ever wants to actually "undo" to it, it's part of the live "undo stack" memory.
 *
 * NOTE 6
 *      The builtin tools use two kinds of shared memory
 *      1) Tool Registers - 8 total, 16bit unsigned integers to allow tools to pass data outside of their own bank
 *          to your plugin
 *      2) Tool Memory    - 16 total, 16bit unsigned integers used to allow tools to pass persistent memory (e.g.
 *          last used brush) settings outside their bank/change tool defaults at startup.  (currently only 9/16 of
 *          the memory locations are used)
 *
 * NOTE 7
 *      The mouse is stored in a number of "precomputed" forms, since it's used in many places like this - it
 *      speeds up the app to cache the sums - here are the versions found here:
 *          mouseX, mouseY              uInt16          raw coorddinate data
 *          canvasX, canvasY            uInt8           data precalcuated to fit within 48K L2 bounds
 *
 * NOTE 8
 *      The Palette Flags memory is a bit-field packed series of flags, currently implemented ones are
 *          Bit 0:           Is palette "modified" from hardware default?
 *
 */

#endif //PLOTIT_API_H
